### PR TITLE
time: lazy init TimerShared  in TimerEntry

### PR DESCRIPTION
**Pull request by wathenjiang**
This commit is part of reducing timeout performance overhead.

See https://github.com/tokio-rs/tokio/issues/6504

Below are relevant benchmark results of this PR on m1:

```
single_thread_timeout   time:   [21.869 ns 21.987 ns 22.135 ns]
                        change: [-3.4429% -2.0709% -0.8759%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

multi_thread_timeout-8  time:   [4.4835 ns 4.6138 ns 4.7614 ns]
                        change: [-4.3554% +0.1643% +4.5114%] (p = 0.95 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  8 (8.00%) high mild
  1 (1.00%) high severe
```

Below are relevant benchmark results of the current version on m1:

```
single_thread_timeout   time:   [40.227 ns 40.416 ns 40.691 ns]
                        change: [+81.321% +82.817% +84.121%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  3 (3.00%) high mild
  11 (11.00%) high severe

multi_thread_timeout-8  time:   [183.16 ns 186.02 ns 188.21 ns]
                        change: [+3765.0% +3880.4% +3987.4%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) low severe
  6 (6.00%) low mild
```

### CODE REVIEW

**Comment by wathenjiang** [comment]
Below are relevant benchmark results of this PR on Linux AMD64:

```
single_thread_timeout   time:   [59.230 ns 59.308 ns 59.389 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

multi_thread_timeout-8  time:   [94.476 ns 95.247 ns 95.877 ns]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
```


Below are relevant benchmark results of the current version on Linux AMD64:

```
single_thread_timeout   time:   [106.43 ns 106.69 ns 107.01 ns]
                        change: [+80.061% +81.176% +82.266%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 16 outliers among 100 measurements (16.00%)
  3 (3.00%) high mild
  13 (13.00%) high severe

multi_thread_timeout-8  time:   [181.75 ns 189.82 ns 199.37 ns]
                        change: [+89.155% +96.135% +103.05%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 19 outliers among 100 measurements (19.00%)
  1 (1.00%) high mild
  18 (18.00%) high severe
```

**Comment by wathenjiang** [comment]
Should we lazily initialize the `TimerEntry` within the `Sleep` instead of `Sleep` within `Timeout`?

**Comment by Darksonn** [comment]
Lazily initializing the `TimerEntry` instead sounds like a good idea.

File: benches/Cargo.toml [hunk 1]
```diff
@@ -90,3 +90,8 @@ harness = false
 name = "time_now"
 path = "time_now.rs"
 harness = false
+
+[[bench]]
+name = "time_timeout"
+path = "time_timeout.rs"
+harness = false
```
Comment (Darksonn) [inline benches/Cargo.toml:97]: This file seems to be missing a newline at the end.

File: benches/time_timeout.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use tokio::time::timeout;
+
+// a vevry quick async task, but might timeout
+async fn quick_job() -> usize {
+    1
+}
+
+fn single_thread_scheduler_timeout(c: &mut Criterion) {
+    do_test(c, 1, "single_thread_timeout");
+}
```
Comment (Darksonn) [inline benches/time_timeout.rs:17]: This is using a multi-threaded runtime with one thread. It would also be nice to include the current-thread runtime.

**Comment by wathenjiang** [comment]
> Lazily initializing the `TimerEntry` instead sounds like a good idea.

I think so. After some time, I find there are many custom `timeout`s in tokio, such as `Timeout<S>` in the `tokio-stream/src/stream_ext/timeout.rs`.

File: benches/Cargo.toml (hunk 1 above)
Comment (wathenjiang) [reply benches/Cargo.toml]: done.

File: tokio/src/time/sleep.rs [hunk 3]
```diff
@@ -412,18 +448,34 @@ impl Sleep {
...
+    // it can avoid the unnecessary creation and drop of `TimerEntry`.
+    fn lazy_init_timer_entry(self: Pin<&mut Self>, deadline: Instant) {
+        let mut me = self.project();
+        *me.deadline = deadline;
+        if me.entry.is_none() {
+            let entry = TimerEntry::new(me.handle, deadline);
+            me.entry.as_mut().set(Some(entry));
+        }
+    }
 }
```
Comment (paolobarbolini) [inline tokio/src/time/sleep.rs:479]: Drive-by review: could this be simplified with [`Option::get_or_insert_with`](https://doc.rust-lang.org/1.63.0/std/option/enum.Option.html#method.get_or_insert_with)?

File: tokio/src/time/sleep.rs [hunk 4]
```diff
@@ -226,10 +227,11 @@ pin_project! {
...
     pub struct Sleep {
         inner: Inner,
-
+        deadline : Instant,
+        handle: scheduler::Handle,
         // The link between the `Sleep` instance and the timer that drives it.
         #[pin]
-        entry: TimerEntry,
+        entry: Option<TimerEntry>,
     }
```
Comment (Darksonn) [inline tokio/src/time/sleep.rs:235]: How does this change the size of `Sleep`? Could `TimerEntry` be changed to reduce the change?

Comment (wathenjiang) [reply tokio/src/time/sleep.rs:235]: I think this should be possible. Your idea looks better, I'll give it a try.

Comment (wathenjiang) [reply tokio/src/time/sleep.rs:235]: I put `deadline` and `handle` in `Sleep` here for do not init the `TimerEntry`, but I have to put them to a place to save them temporarily.

Saving `handle` here is to ensure the `timeout` will panic if there is no runtime handle in TLS or we do not `enable_time` or `enable_all`.

The size of `Sleep` is an issue worth considering. Is there any way to solve this problem?

**Comment by wathenjiang** [comment]
The below are benchmark where the timer always gets registered with the driver. 

```
single_thread_timeout   time:   [37.361 ns 37.515 ns 37.723 ns]
                        change: [-5.5281% -3.6420% -1.9411%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  5 (5.00%) high mild
  13 (13.00%) high severe

multi_thread_timeout-8  time:   [158.60 ns 162.22 ns 165.25 ns]
                        change: [+1.5142% +3.7530% +5.9648%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) low severe
  5 (5.00%) low mild

single_thread_sleep     time:   [37.637 ns 37.746 ns 37.916 ns]
                        change: [-2.4402% -1.0747% -0.1511%] (p = 0.05 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

multi_thread_sleep-8    time:   [151.48 ns 155.76 ns 159.73 ns]
                        change: [-4.1891% -1.8008% +0.8186%] (p = 0.17 > 0.05)
                        No change in performance detected.
```

In fact, the `change` column in the previous test result is relative to this, such as `change: [+9.1588% +10.095% +11.148%] (p = 0.00 < 0.05)` in `single_thread_timeout`.

**Comment by wathenjiang** [comment]
Of course, if all timers will be registered in the Driver, and the concurrency is very high. Then lock contention will become the most important concern.

The following flame shows the global lock ops take `7.15%` in `register`, and `8.75%` in `clear_entry`. 

image
The benchmark test code:

```rust
fn main() {
    let runtime = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .worker_threads(8)
        .build()
        .unwrap();

    let _r = runtime.block_on(async {
        let mut handles = Vec::with_capacity(1024);
        for _ in 0..1024 {
            handles.push(tokio::spawn(async move {
                loop {
                    for i in 1..10 {
                        for _ in 0..100 {
                            let h = timeout(Duration::from_millis(i), never_ready_job2(i));
                            let _r = h.await;
                        }
                    }
                }
            }));
        }
        for handle in handles {
            handle.await.unwrap();
        }
    });
}

pub fn fibonacci(n: u64) -> u64 {
    match n {
        0 => 1,
        1 => 1,
        n => fibonacci(n - 1) + fibonacci(n - 2),
    }
}

// a job never ready
async fn never_ready_job2(n : u64) -> u64 {
    let res = fibonacci(n+5); // do some logic
    pending::<()>().await;
    res
}
```

The global lock in multi-threads is usually not good enough in performance.

So, going back to the issue itself, both methods may need to be done to reduce the performance overhead of timeout.

- In most cases where `timeout` is not registered with the driver, lazy initialization optimization is used
- Most `timeout`s need to be registered to the driver, using a mechanism similar to shard locks

File: tokio/src/time/sleep.rs (hunk 4 above)
Comment (wathenjiang) [reply tokio/src/time/sleep.rs:235]: In my test, it can improve the performance by reduicng lock contention like lazy initialization when we always let timeout register into the driver. But, on the other hand, if the timeout never register into the driver, this can not bring significant performance improvement.

Comment (wathenjiang) [reply tokio/src/time/sleep.rs:235]: That said, maybe we can use `inner: StdUnsafeCell<Option<TimerShared>>` for this purpose, the size of `Sleep` should be barely changed.

Avoiding `clear_entry` for unregistered `timeout`s is not necessary. In this case, we should avoid constructing the intrusive linked list item.

File: tokio/src/time/sleep.rs (hunk 3 above)
Comment (wathenjiang) [reply tokio/src/time/sleep.rs]: Thanks!

**Comment by wathenjiang** [comment]
By lazily initializing the `TimerShared` in `TimerEntry`, we get a performance improvement. The below are the benchmark results (On Linux AMD64 which has 16 cores ):

The results of this PR:
```
single_thread_timeout   time:   [72.079 ns 72.146 ns 72.220 ns]
multi_thread_timeout-8  time:   [104.14 ns 104.90 ns 105.49 ns]
single_thread_sleep     time:   [69.147 ns 69.344 ns 69.592 ns]
multi_thread_sleep-8    time:   [106.08 ns 106.60 ns 106.98 ns]
```
The results of master:

```
single_thread_timeout   time:   [107.02 ns 107.24 ns 107.52 ns]
multi_thread_timeout-8  time:   [195.11 ns 201.26 ns 208.99 ns]
single_thread_sleep     time:   [116.30 ns 116.44 ns 116.61 ns]
multi_thread_sleep-8    time:   [210.79 ns 223.30 ns 236.26 ns]
```

To achieve all of this, our price is that the byte size of `Sleep` has increased from 112 to 120.

File: benches/time_timeout.rs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+// a vevry quick async task, but might timeout
+async fn quick_job() -> usize {
+    1
+}
+
+fn build_run_time(workers: usize) -> Runtime {
+    if workers == 1 {
+        tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .worker_threads(workers)
```
Comment (mox692) [inline benches/time_timeout.rs:22]: I think if we use the current_thread runtime, we wouldn't need to specify the number of worker threads.

File: benches/time_timeout.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
+//! Benchmark spawning a task onto the basic and threaded Tokio executors.
+//! This essentially measure the time to enqueue a task in the local and remote
+//! case.
```
Comment (mox692) [inline benches/time_timeout.rs:3]: Doesn't seem to be the text describing this benchmark? I guess it could be revised or removed.

**Review by Darksonn — COMMENTED** [review]
Apparently I didn't actually submit this comment. It's been "pending" for 5 days. Sorry about that.

File: benches/time_timeout.rs (hunk 5 above)
Comment (wathenjiang) [reply benches/time_timeout.rs:22]: Yes, I will delete this.

File: benches/time_timeout.rs (hunk 6 above)
Comment (wathenjiang) [reply benches/time_timeout.rs:3]: It is copied from the benchmark in `time.rs`. I think deleting this comment is acceptable for this benchmark.

File: benches/time_timeout.rs (hunk 5 above)
Comment (wathenjiang) [reply benches/time_timeout.rs]: Thanks!

File: tokio/src/runtime/time/entry.rs [hunk 7]
```diff
@@ -477,35 +477,42 @@ unsafe impl linked_list::Link for TimerShared {
...
 
+    fn is_inner_init(&self) -> bool {
+        unsafe { &*self.inner.get() }.is_some()
+    }
+
+    // This lazy initialization is for performance purposes.
     fn inner(&self) -> &TimerShared {
-        unsafe { &*self.inner.get() }
+        unsafe { &mut *self.inner.get() }.get_or_insert(TimerShared::new())
     }
```
[the remark this answers is not in the document]
Comment (wathenjiang) [reply tokio/src/runtime/time/entry.rs]: I have used additional `is_done` code  to achieve this goal.

File: benches/time_timeout.rs [hunk 8]
```diff
@@ -0,0 +1,109 @@
+use std::time::{Duration, Instant};
+
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use tokio::{
+    runtime::Runtime,
+    time::{sleep, timeout},
+};
+
+// a vevry quick async task, but might timeout
```
Comment (FrankReh) [inline benches/time_timeout.rs:9]: "very"

Comment (wathenjiang) [reply benches/time_timeout.rs:9]: Thanks.

**Review by Darksonn — APPROVED** [review]
Looks good to me.
